### PR TITLE
Name resolution: resolve interfaces in expressions

### PR DESCRIPTION
**Pull request by auduchinok**
Here's an experiment to see what breaks in CI.

### CODE REVIEW

**Comment by T-Gro** [comment]
Is this still an experiment, or ready for review?

I have a feeling that the typecheck error reported now is worse from understandability points of view => I would need convincing that the tradeoff is worth it.

**Comment by auduchinok** [comment]
@T-Gro This is currently still an experiment. I'm currently on vacation and going to get back to it.

The general idea here is to provide symbols (with the type arguments where possible) and have better errors for unfinished expressions like:
```fsharp
// module/class/record/union
List
List.

// interface
IList
IList<_>
IList<int>
```

Currently no symbols are provided and the errors are misleading. This experiment already improves things for some of these cases, but needs further work.

The change in the tests is mostly about quite a corner case related to incorrect type definitions, and the previous error about constructors is also misleading. I'm going to look into improving it; maybe I'll just try to reuse this error.

**Comment by auduchinok** [comment]
> I have a feeling that the typecheck error reported now is worse from understandability points of view => I would need convincing that the tradeoff is worth it.

@T-Gro I think they're actually better. The most of the updated baselines capture cases where there's a type that can't be instantiated properly due to an issue in its declaration, and previous errors would say that the types are undefined (which is not true and is misleading) while the newer errors say that the types are used in a wrong way, which is probably good enough considering that a the fix is needed on the declaration side, and there's an appropriate error reported for the declaration.

**Comment by auduchinok** [comment]
This is ready. It changes the way how wrong usages of interfaces are encoded, but it looks good and makes interface items are more inline with rest of the types.

File: src/Compiler/Service/ServiceDeclarationLists.fs [hunk 1]
```diff
@@ -1107,9 +1089,9 @@ type DeclarationListInfo(declarations: DeclarationListItem[], isForType: bool, i
             items 
             |> List.map (fun x ->
                 match x.Item with
+                | Item.Types (_, TType_app(tcref, _, _) :: _) when isInterfaceTyconRef tcref -> { x with MinorPriority = 1000 + tcref.TyparsNoRange.Length }
```
Comment (psfinaki) [inline src/Compiler/Service/ServiceDeclarationLists.fs:1092]: For my understanding, what does this change mean here - why was it needed?

File: src/Compiler/Checking/NameResolution.fsi [hunk 2]
```diff
@@ -385,6 +382,7 @@ type internal ItemOccurence =
     | Implemented
     | RelatedText
     | Open
+    | WrongUse
```
Comment (psfinaki) [inline src/Compiler/Checking/NameResolution.fsi:385]: InvalidUse?

File: src/Compiler/Checking/NameResolution.fs [hunk 3]
```diff
@@ -276,8 +273,7 @@ type Item =
             | ValueSome tcref -> tcref.DisplayNameCore
             | _ -> nm
             |> DemangleGenericTypeName
-        | Item.CtorGroup(nm, _) -> nm |> DemangleGenericTypeName
-        | Item.FakeInterfaceCtor ty
+        |Item.CtorGroup(nm, _) -> nm |> DemangleGenericTypeName
```
Comment (psfinaki) [inline src/Compiler/Checking/NameResolution.fs:276]:
```suggestion
        | Item.CtorGroup(nm, _) -> nm |> DemangleGenericTypeName 
```

**Review by psfinaki — APPROVED** [review]
Mhm I guess the resulting message could be clearer but IMO the new way it is not worse than what used to be. Okay with me.

File: src/Compiler/Checking/NameResolution.fsi (hunk 2 above)
Comment (auduchinok) [reply src/Compiler/Checking/NameResolution.fsi]: Sounds good, thanks.

File: src/Compiler/Service/ServiceDeclarationLists.fs (hunk 1 above)
Comment (auduchinok) [reply src/Compiler/Service/ServiceDeclarationLists.fs:1092]: That's just a rewrite of the removed pattern below.

**Review by T-Gro — APPROVED** [review]

**Comment by auduchinok** [comment]
This is green now.
